sensor_test.py

```diff
@@ -0,0 +1,240 @@
+
+    print("Request firmware update")
+    update_firmware()
+    wait(
+        func=get_sensor_info,
+        condition=lambda x: isinstance(x, dict),
+        tries=5,
+        timeout=3,
+    )
+
+    print("Get current sensor firmware version")
+    updated_sensor_info = get_sensor_info()
+    current_firmware_version = updated_sensor_info.get("firmware_version")
+
+    print("Validate that current firmware version is +1 to original firmware version")
+    assert (
+        current_firmware_version == original_firmware_version + 1
+    ), "Sensor firmware was not updated"
+
+    print("Repeat steps 1-4 until sensor is at max_firmware_version - 1")
+    max_firmware_version = 15
+    while True:
+        update_firmware()
+        wait(
+            func=get_sensor_info,
+            condition=lambda x: isinstance(x, dict),
+            tries=5,
+            timeout=3,
+        )
+        updated_sensor_info = get_sensor_info()
+        current_firmware_version = updated_sensor_info.get("firmware_version")
+        if current_firmware_version == max_firmware_version - 1:
+            break
+
+    print("Update sensor to max firmware version")
+    while True:
+        update_firmware()
+        wait(
+            func=get_sensor_info,
+            condition=lambda x: isinstance(x, dict),
+            tries=5,
+            timeout=3,
+        )
+        updated_sensor_info = get_sensor_info()
+        current_firmware_version = updated_sensor_info.get("firmware_version")
+        if current_firmware_version == max_firmware_version:
+            break
+
+    print("Validate that sensor is at max firmware version")
+    sensor_info = get_sensor_info()
+    current_firmware_version = sensor_info.get("firmware_version")
+    assert (
+        current_firmware_version == max_firmware_version
+    ), "Sensor firmware version is not at max possible"
+
+    print("Request another firmware update")
+    sensor_info = get_sensor_info()
+    firmware_version_before_update = sensor_info.get("firmware_version")
+
+    update_sensor_firmware = update_firmware()
+
+    wait(
+        func=get_sensor_info,
+        condition=lambda x: isinstance(x, dict),
+        tries=5,
+        timeout=3,
+    )
+
+    print("Validate that sensor doesn't update and responds appropriately")
+
+    sensor_info = get_sensor_info()
+    firmware_version_after_update = sensor_info.get("firmware_version")
+
+    assert (
+        update_sensor_firmware == "already at latest firmware version"
+        and firmware_version_after_update == firmware_version_before_update
+    ), "Something went wrong and firmware version was updated unexpectedly"
+
+    print(
+        "Validate that sensor firmware version doesn't change if it's at maximum value"
+    )
+    sensor_info = get_sensor_info()
+    current_firmware_version = sensor_info.get("firmware_version")
+    if current_firmware_version == max_firmware_version:
```

Comment:
Code after line 188 is redundant because you've already checked that the sensor is at max version and that it gives a proper response on line 186.

As I mentioned in one of the comments above, see if you can restructure this test to be just two things: a loop updating the sensor to the maximum allowed version and a block validating that it's not possible to update any further.